langchain/retrievers/multi_query.py

```diff
@@ -0,0 +1,148 @@
+        By generating multiple perspectives on the user question, 
+        your goal is to help the user overcome some of the limitations 
+        of distance-based similarity search. Provide these alternative 
+        questions seperated by newlines. Original question: {question}""",
+    )
+
+    def __init__(
+        self,
+        retriever: BaseRetriever,
+        num_queries: int,
+        llm_chain: LLMChain,
+        verbose: bool = True,
+    ) -> None:
+        """Initialize MultiQueryRetriever.
+
+        Args:
+            retriever: retriever to query documents from
+            num_queries: number of queries for the LLM to generate
+            llm_chain: llm_chain for query generation
+            verbose: show the queries that we generated to the user
+
+        Returns:
+            MultiQueryRetriever
+        """
+        self.retriever = retriever
+        self.num_queries = num_queries
+        self.llm_chain = llm_chain
+        self.verbose = verbose
+
+    @classmethod
+    def from_llm(
+        cls,
+        retriever: BaseRetriever,
+        num_queries: int,
+        llm: BaseLLM,
+        prompt: PromptTemplate = DEFAULT_QUERY_PROMPT,
+    ) -> "MultiQueryRetriever":
+        """Initialize from llm using default template.
+
+        Args:
+            retriever: retriever to query documents from
+            num_queries: number of queries for the LLM to generate
+            llm: llm for query generation using DEFAULT_QUERY_PROMPT
+
+        Returns:
+            MultiQueryRetriever
+        """
+        output_parser = LineListOutputParser()
+        llm_chain = LLMChain(llm=llm, prompt=prompt, output_parser=output_parser)
+        return cls(retriever=retriever, num_queries=num_queries, llm_chain=llm_chain)
+
+    def get_relevant_documents(self, question: str) -> List[Document]:
+        """Get relevated documents given a user query.
+
+        Args:
+            question: user query
+
+        Returns:
+            Unique union of relevant documents from all generated queries
+        """
+        queries = self.generate_queries(question)
+        documents = self.retrieve_documents(queries)
+        unique_documents = self.unique_union(documents)
+        return unique_documents
+
+    async def aget_relevant_documents(self, query: str) -> List[Document]:
+        raise NotImplementedError
+
+    def generate_queries(self, question: str) -> List[str]:
+        """Generate queries based upon user input.
+
+        Args:
+            question: user query
+
+        Returns:
+            List of LLM generated queries that are similar to the user input
+        """
+        response = self.llm_chain(
+            {"question": question, "num_queries": self.num_queries}
+        )
+        if self.verbose:
+            # This hard-coding of lines will be probalematic
+            print(f"Generated queries: {response['text'].lines}")
+        return response["text"].lines
```

Comment:
This is brittle. Need a better solution, since a user can pass an arbitrary output parser.

Reply:
One option is passing the `parser_key` as a parameter. Pushes more responsibility onto the user, but also affords the flexibility to supply your own prompt / output parser. Of course, it can be run in "easy mode" with these abstracted away from the user, as shown in the notebook.
```
llm = ChatOpenAI(temperature=0)
retriever_from_llm = MultiQueryRetriever.from_llm(retriever=vectordb.as_retriever(),num_queries=num_queries,llm=llm)
```